### DISCUSSION TITLE
Roadmapping: Graceful shutdown of managed processes

### DISCUSSION THREAD

**Discussion by robgjansen** [shadow/shadow#2034] [Show and tell]
Plan:
- defer graceful shutdown until either killall or fork/exec is implemented.
- We'll document PID numbering in the meantime as a workaround

Reasoning:
- Could implement killall, but that requires /proc and we want to support a virtual filesystem framework for /proc which is considerably more work. The earliest we'd want to consider working on the virtual filesystem is v2.3 (v2.2 is for rust design work).
- Could implement fork/exec to allow simple bash wrappers when launching programs. But we are going to defer that until v2.3 at the earliest.

**Comment by sporksmith** [comment]
Documented PID numbering, and closed #1491 since that enables graceful shutdown.

Supporting killall workflow tracked in #1986 

Supporting fork+exec tracked in #1987

**Reply by stevenengler** [reply]
Related issue: "Shadow will return success even when plugins are killed by Linux OOM killer" — #2151

Related issue: "Output files are empty for processes that don't exit" — #2597

**Comment by stevenengler** [comment]
I think Shadow's current behaviour is confusing, and we've discussed how we might want this to work in the future.

Generally when a managed process exits, Shadow will check its exit code and if the exit code is non-zero, the simulation will eventually be marked as "failed". Currently when the simulation ends, Shadow will send a SIGKILL to all processes that are still running. These processes that are killed at the end of the simulation will not affect the successful/failed status of the simulation. This is sometimes what the user wants, for example long-running processes like tor or web servers will typically run until the simulation ends. But sometimes this is not what the user wants, for example if a process like curl is unable to connect to the server, continues trying until the end of the simulation, and is never successful.

Once we have better support for shutting down managed processes, we can make it so that processes killed at the end of the simulation *do* affect the successful/failed status of the simulation. This would be a breaking change, so we'd have to do this in a major version bump. For applications like tor or web servers, the user would need to kill them explicitly with for example `kill -sigterm 1001` and have the programs exit with a zero error code.

**Comment by stevenengler** [comment]
We discussed this some more and haven't yet decided on a path forward. There are a few independent questions:

1. Do users expect Shadow to return an error status if a managed process exits with a non-zero error code? What about when Shadow intentionally kills a managed process due to `stop_time`?

2. What should Shadow's default behaviour be? Should Shadow even care about the exit codes of managed processes? If a managed process exits with a non-zero exit code, but Shadow itself didn't have any errors, then should Shadow return a zero or non-zero exit code? In this case Shadow executed everything as specified by the user so the simulation itself didn't have any errors, but the simulation results may not be what the user expected. If we were to classify these as "logical errors" or "technical errors" in the simulation, which of these two should affect the simulation's exit code?

3. How can we be more clear to the user about the end result of a simulation, especially if the answer to (1) conflicts with the answer to (2)? Currently when Shadow finishes, it writes a "Shadow completed successfully" or "Shadow did not complete successfully" message depending on Shadow's exit code.

4. Sometimes the "user" cannot read and understand Shadow's log messages to look for logical errors in their simulation. For example if we want to make sure that our tests are working correctly in the CI, we need a way to automate checking the output from (3). If the answer to (2) is to ignore exit codes, we still need a way to check these exit codes in our tests.

5. Should the configuration format support options to change how Shadow handles different managed process exit states? Or can we wait until Shadow supports fork/exec, and require users to wrap managed processes in bash scripts which check exit states?

------------

From discussion:

1. We don't know exactly what users expect from Shadow, but they probably expect some consistency about when Shadow returns a non-zero exit code. Shadow's current rules about when to return a non-zero exit code are confusing and not documented anywhere.

2. We have conflicting opinions on this.

3. Instead of printing a success/failure message, we can print a summary table of processes and their exit states. There are different ways we could summarize these states, but one would be to group them based on (plugin_path, exit_state). We should also document Shadow's behaviour somewhere.

4. We could log exit codes to a json file, and have a script iterate over that json to look for managed processes that didn't return a zero exit code. But this probably isn't ideal.

5. We could have a per-process configuration option that changes how Shadow handles different managed process exit states. The default for this option would depend on the answer to (2). A backwards-compatible option could be a "runs_forever" option that has the following rules:
    - if false: It is an error if the plugin exits with a non-zero error code, either during the simulation, or at the end of the simulation after the sigkill. If it's running at the end of the simulation, it will always be an error.
    - if true: It is an error if the plugin exits before the end of the simulation. The exit code is always ignored (both during and at the end of the simulation).

    This might still not be clear to users though. Another option could be to have a "exit_state" option that allows users to specify exactly what exit state the process should have (an exit code, an exit signal, or if it should be killed by Shadow's "stop time").

**Reply by jtracey** [reply]
An idle thought: perhaps use the exit_state option, and then have two default configs a user must choose from at runtime: a "testing" config, intended for integrating into other applications' tests and CI workflows, sanity checking experiment configs, etc., that has defaults optimized for such (coarser sim time granularity for better performance, exit_state set to return exit codes, etc.), and a "simulation" config, intended for higher-accuracy, more precise experiments (similar to current config).

**Comment by robgjansen** [comment]
We discussed this again at another roadmapping session and came with the following plan.

- proposed plan:
  - shadow should return an error if process exits with non-zero exit code
  - shadow should return an error if process is still running at end
  - shadow sends a signal at time `stop_time` (rename `shutdown_time`?)
  - by default, that signal is SIGTERM, but overridable (`shutdown_signal`)
  - shadow lets process keep running after sending the stop/shutdown signal, up until global simulation stop time
  - user can override expected end state from "exited with code 0" to "exited with code X", "still running", or "don't care"
    - `expected_end_state` = Enum(DontCare | Exited(i32) | Signaled(i32) | StillRunning)
- better logging at end of the simulation
- error-level log messages print to stderr
- reduce number of error level logs (for example don't error-log on ENOSYS)
- continue counting plugin errors and print at the end of the simulation
  - Note: requires v3.0, and we probably want to update tornettools
